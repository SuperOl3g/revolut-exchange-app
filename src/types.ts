import React from 'react';
import { ThunkDispatch } from 'redux-thunk';
import { IStoreState } from './store';
import { ActionType } from './actions';

export type TCurrency = 'RUB' | 'USD' | 'EUR' | 'GBP';

export type TDispatch = ThunkDispatch<IStoreState, undefined, IAction>;

export interface IAction {
  type: ActionType;
  payload: {
    [key: string]: any;
  };
}

export interface IFieldCallback {
  (
    e: React.SyntheticEvent<HTMLElement> | null,
    obj: {
      value: string;
      name?: string;
    }
  ): void;
}

export enum FieldType {
  Source,
  Target
}

export interface IAlert {
  id: string;
  message: string;
  repeatable?: boolean;
}

export type TAnyObject = {
  [id in string | number]?: any;
};
