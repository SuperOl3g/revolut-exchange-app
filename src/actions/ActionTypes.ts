export enum ActionType {
  RATES_REQUEST = 'RATES_REQUEST',
  RATES_REQUEST_SUCCESS = 'RATES_REQUEST_SUCCESS',

  SOURCE_CURRENCY_CHANGE = 'SOURCE_CURRENCY_CHANGE',
  TARGET_CURRENCY_CHANGE = 'TARGET_CURRENCY_CHANGE'
}
