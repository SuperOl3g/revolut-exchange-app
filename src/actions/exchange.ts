import { IAction, TCurrency } from '../types';
import { ActionType } from './index';

export default function exchange(
  sourceCurrency: TCurrency,
  sourceAmount: number,
  targetCurrency: TCurrency,
  targetAmount: number
): IAction {
  return {
    type: ActionType.POCKETS_EXCHANGE,
    payload: {
      sourceCurrency,
      sourceAmount,
      targetCurrency,
      targetAmount
    }
  };
}
