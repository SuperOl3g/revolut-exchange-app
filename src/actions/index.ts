export enum ActionType {
  RATES_REQUEST = 'RATES_REQUEST',
  RATES_REQUEST_SUCCESS = 'RATES_REQUEST_SUCCESS',

  SOURCE_CURRENCY_CHANGE = 'SOURCE_CURRENCY_CHANGE',
  TARGET_CURRENCY_CHANGE = 'TARGET_CURRENCY_CHANGE',

  POCKETS_EXCHANGE = 'POCKETS_EXCHANGE',

  ALERT_ADD = 'ALERT_ADD',
  ALERT_REMOVE = 'ALERT_REMOVE'
}
