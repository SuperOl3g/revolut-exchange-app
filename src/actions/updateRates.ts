import { Dispatch } from 'redux';
import { EXCHANGE_API } from '../constants/common';
import { ActionType } from './ActionTypes';
import { IStoreState } from '../store';

export default function updateRates() {
  return (dispatch: Dispatch, getState: () => IStoreState) => {
    const { sourceCurrency } = getState().exchange;

    dispatch({
      type: ActionType.RATES_REQUEST,
      payload: { baseCurrency: sourceCurrency }
    });

    return fetch(`${EXCHANGE_API}/latest?base=${sourceCurrency}`)
      .then(response => response.json())
      .then(({ rates }) => {
        const { RUB, EUR, USD, GBP } = rates;

        return { RUB, EUR, USD, GBP };
      })
      .then(rates =>
        dispatch({
          type: ActionType.RATES_REQUEST_SUCCESS,
          payload: { rates, baseCurrency: sourceCurrency }
        })
      );
  };
}
