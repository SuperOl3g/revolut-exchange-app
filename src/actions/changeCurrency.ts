import { ActionType } from './index';
import { FieldType, TCurrency, TDispatch } from '../types';
import { IStoreState } from '../store';
import updateRates from './updateRates';
import { DEFAULT_CURRENCY_ORDER } from '../constants/common';

function getNextCurrency(currency: TCurrency): TCurrency {
  return DEFAULT_CURRENCY_ORDER[
    (DEFAULT_CURRENCY_ORDER.indexOf(currency) + 1) %
      DEFAULT_CURRENCY_ORDER.length
  ];
}

export default function changeCurrency(type: FieldType, currency: TCurrency) {
  return (dispatch: TDispatch, getState: () => IStoreState) => {
    dispatch({
      type:
        type === FieldType.Source
          ? ActionType.SOURCE_CURRENCY_CHANGE
          : ActionType.TARGET_CURRENCY_CHANGE,
      payload: { currency }
    });

    dispatch(updateRates());

    const state = getState();
    const oppositeCurrency =
      state.exchange[
        type === FieldType.Source ? 'targetCurrency' : 'sourceCurrency'
      ];

    // if currencies are same we need to change one of them
    if (oppositeCurrency === currency) {
      dispatch({
        type:
          type === FieldType.Source
            ? ActionType.TARGET_CURRENCY_CHANGE
            : ActionType.SOURCE_CURRENCY_CHANGE,
        payload: { currency: getNextCurrency(oppositeCurrency) }
      });
    }
  };
}
