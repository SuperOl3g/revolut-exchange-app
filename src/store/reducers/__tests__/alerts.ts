import reducer, { initState } from '../alerts';
import { ActionType } from '../../../actions';
import { IAction } from '../../../types';

describe('Alerts reducer', () => {
  it('should return the initial state', () => {
    expect(reducer(undefined, {} as IAction)).toMatchSnapshot();
  });

  it('should add new alert', () => {
    expect(
      reducer(initState, {
        type: ActionType.ALERT_ADD,
        payload: {
          id: 'MY_ALERT',
          message: 'Some text'
        }
      })
    ).toMatchSnapshot();
  });

  it('should renew unrepeatable alerts', () => {
    const state = {
      0: { id: 'ALERT0', message: 'text0' },
      1: { id: 'ALERT1', message: 'text1' },
      2: { id: 'ALERT2', message: 'text2' },
      3: { id: 'ALERT3', message: 'text3' }
    };

    expect(
      reducer(state, {
        type: ActionType.ALERT_ADD,
        payload: {
          id: 'ALERT2',
          message: 'Some text',
          repeatable: false
        }
      })
    ).toMatchSnapshot();
  });

  it('should remove old alert', () => {
    const state = {
      0: { id: 'ALERT0', message: 'text0' },
      1: { id: 'ALERT1', message: 'text1' },
      2: { id: 'ALERT2', message: 'text2' },
      3: { id: 'ALERT3', message: 'text3' }
    };

    expect(
      reducer(state, {
        type: ActionType.ALERT_REMOVE,
        payload: {
          id: 'ALERT1',
          message: 'Some text'
        }
      })
    ).toMatchSnapshot();
  });
});
