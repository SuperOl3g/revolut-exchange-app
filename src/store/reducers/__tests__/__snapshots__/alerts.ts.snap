// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alerts reducer should add new alert 1`] = `
Object {
  "0": Object {
    "id": "MY_ALERT",
    "message": "Some text",
  },
}
`;

exports[`Alerts reducer should remove old alert 1`] = `
Object {
  "0": Object {
    "id": "ALERT0",
    "message": "text0",
  },
  "2": Object {
    "id": "ALERT2",
    "message": "text2",
  },
  "3": Object {
    "id": "ALERT3",
    "message": "text3",
  },
}
`;

exports[`Alerts reducer should renew unrepeatable alerts 1`] = `
Object {
  "0": Object {
    "id": "ALERT0",
    "message": "text0",
  },
  "1": Object {
    "id": "ALERT1",
    "message": "text1",
  },
  "3": Object {
    "id": "ALERT3",
    "message": "text3",
  },
  "4": Object {
    "id": "ALERT2",
    "message": "Some text",
    "repeatable": false,
  },
}
`;

exports[`Alerts reducer should return the initial state 1`] = `Object {}`;
